File: app/src/main/java/com/nytimes/android/sample/activity/PersistingStoreActivity.java
package com.nytimes.android.sample.activity;

import android.support.v7.app.AppCompatActivity;


public class PersistingStoreActivity extends AppCompatActivity {

//    private RecyclerView recyclerView;
//    private PostAdapter postAdapter;
//    private Store<RedditData, BarCode> persistedStore;
//
//    @Override
//    protected void onCreate(Bundle savedInstanceState) {
//        super.onCreate(savedInstanceState);
//
//        setContentView(R.layout.activity_store);
//        setSupportActionBar((Toolbar) findViewById(R.id.toolbar));
//
//        postAdapter = new PostAdapter();
//        recyclerView = (RecyclerView) findViewById(R.id.postRecyclerView);
//        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
//        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
//        recyclerView.setLayoutManager(layoutManager);
//        recyclerView.setAdapter(postAdapter);
//    }
//
//    private void initStore() {
//        if (this.persistedStore == null) {
//            this.persistedStore = ((SampleApp) getApplicationContext()).getPersistedStore();
//        }
//    }
//
//    @SuppressWarnings("CheckReturnValue")
//    public void loadPosts() {
//        BarCode awwRequest = new BarCode(RedditData.class.getSimpleName(), "aww");
//
//        /*
//        First call to get(awwRequest) will use the network, then save response in the in-memory
//        cache. Subsequent calls will retrieve the cached version of the data.
//         */
//        this.persistedStore
//                .get(awwRequest)
//                .flatMapObservable(this::sanitizeData)
//                .toList()
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe(this::showPosts, throwable ->
//                        Log.e(StoreActivity.class.getSimpleName(), throwable.getMessage(),
//                                throwable));
//    }
//
//    private void showPosts(List<Post> posts) {
//        postAdapter.setPosts(posts);
//        makeText(PersistingStoreActivity.this,
//                "Loaded " + posts.size() + " posts",
//                Toast.LENGTH_SHORT)
//                .show();
//    }
//
//    private Observable<Post> sanitizeData(RedditData redditData) {
//        return Observable.fromIterable(redditData.data().children())
//                .map(Children::data);
//    }
//
//    @Override
//    protected void onResume() {
//        super.onResume();
//        initStore();
//        loadPosts();
//    }
}

File: app/src/main/java/com/nytimes/android/sample/PersistingStoreActivity.kt
package com.nytimes.android.sample

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.Toolbar
import android.view.View
import android.widget.Toast
import android.widget.Toast.makeText
import com.nytimes.android.external.store3.base.impl.BarCode
import com.nytimes.android.external.store3.base.impl.Store
import com.nytimes.android.sample.data.model.Children
import com.nytimes.android.sample.data.model.Post
import com.nytimes.android.sample.data.model.RedditData
import com.nytimes.android.sample.reddit.PostAdapter
import com.squareup.moshi.Moshi
import kotlinx.android.synthetic.main.activity_store.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlin.coroutines.CoroutineContext


class PersistingStoreActivity : AppCompatActivity(), CoroutineScope {
    lateinit var job: Job

        override val coroutineContext: CoroutineContext
            get() = job + Dispatchers.Main

    lateinit var postAdapter: PostAdapter
    lateinit var persistedStore: Store<RedditData, BarCode>
    lateinit var moshi: Moshi

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        job = Job()

        setContentView(R.layout.activity_store)
        setSupportActionBar(findViewById<View>(R.id.toolbar) as Toolbar)

        postAdapter = PostAdapter()
        val layoutManager = LinearLayoutManager(this)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        postRecyclerView.layoutManager = layoutManager
        postRecyclerView.adapter = postAdapter
        persistedStore = (applicationContext as SampleApp).persistedStore
        moshi = (applicationContext as SampleApp).moshi
    }

    fun loadPosts() {
        val awwRequest = BarCode(RedditData::class.java.simpleName, "aww")

        /*
        First call to get(awwRequest) will use the network, then save response in the in-memory
        cache. Subsequent calls will retrieve the cached version of the data.
         */
        launch {
            showPosts(
                    sanitizeData(
                            persistedStore.get(awwRequest)))
        }

    }

    private fun showPosts(posts: List<Post>) {
        postAdapter.setPosts(posts)
        makeText(this@PersistingStoreActivity,
                "Loaded ${posts.size} posts",
                Toast.LENGTH_SHORT)
                .show()
    }

    fun sanitizeData(redditData: RedditData): List<Post> =
            redditData.data.children.map(Children::data)

    override fun onResume() {
        super.onResume()
        loadPosts()
    }

    override fun onDestroy() {
        super.onDestroy()
        job.cancel() // all children coroutines gets destroyed automatically
    }
}
